torch/csrc/distributed/c10d/init.cpp

```diff
@@ -973,16 +973,57 @@ that adds a prefix to each key inserted to the store.
 #endif
 
   shared_ptr_class_<::c10d::ProcessGroup::Work>(module, "Work")
-      .def("is_completed", &::c10d::ProcessGroup::Work::isCompleted)
-      .def("is_success", &::c10d::ProcessGroup::Work::isSuccess)
-      .def("exception", &::c10d::ProcessGroup::Work::exception)
-      .def("source_rank", &::c10d::ProcessGroup::Work::sourceRank)
+      .def(
+          "is_completed",
+          [](::c10d::ProcessGroup::Work& work) -> bool {
+            TORCH_WARN_ONCE("ProcessGroup::Work::is_completed API is being "
+                "deprecated, please ping "
+                "https://github.com/pytorch/pytorch/issues/46291 "
+                "if you see this warning");
+            return work.isSuccess();
+          })
+      .def(
+          "is_success",
+          [](::c10d::ProcessGroup::Work& work) -> bool {
+            TORCH_WARN_ONCE("ProcessGroup::Work::is_success API is being "
+                "deprecated, please ping "
+                "https://github.com/pytorch/pytorch/issues/46291 "
+                "if you see this warning");
+            return work.isSuccess();
+          })
+      .def(
+          "exception",
+          [](::c10d::ProcessGroup::Work& work) -> std::exception_ptr {
+            TORCH_WARN_ONCE("ProcessGroup::Work::exception API is being "
+                "deprecated, please ping "
+                "https://github.com/pytorch/pytorch/issues/46291 "
+                "if you see this warning");
+            return work.exception();
+          })
+      .def(
+          "source_rank",
+          [](::c10d::ProcessGroup::Work& work) -> int {
+            TORCH_WARN_ONCE("ProcessGroup::Work::source_rank API is being "
+                "deprecated, please ping "
+                "https://github.com/pytorch/pytorch/issues/46291 "
+                "if you see this warning");
+            return work.sourceRank();
```

Comment:
I don't like preprocessor macro, but it seem to be ideal to avoid string literal duplication, i.e. use something like
```
#define _DEPRECATED_DEF(name, method) \
  def( name, [](::c10d::ProcessGroup::Work& work) -> int { \
  TORCH_WARN_ONCE("ProcessGroup::Work::" name " API is being " \
                                        "deprecated, please ping " \
                                        "https://github.com/pytorch/pytorch/issues/46291 " \
                                        "if you see this warning"); \
   return work.method(); \
   })
```
```suggestion
      ._DEPRECATED_DEF("is_completed", isCompleted)
      ._DEPRECATED_DEF("is_success", isSuccess)
      ._DEPRECATED_DEF("exception", exception)
      ._DEPRECATED_DEF("source_rank", sourceRank)
```

Reply:
Another option is `constexpr` + `fmt::format`

Reply:
added a macro

Reply:
... and switched to fmt::format since clang-tidy doesn't like macros